src/windows/Microsoft.Authentication.Helper.Windows/Application.cs

```diff
@@ -72,18 +61,45 @@ private static string GetArgument(IDictionary<string, string> inputDict, string
 
         protected virtual async Task<string> GetAccessTokenAsync(string authority, string clientId, Uri redirectUri, string resource)
         {
-            var cache = new VisualStudioTokenCache(Context);
-            var authContext = new AuthenticationContext(authority, cache);
+            string[] scopes = { $"{resource}/.default" };
+
+            var appBuilder = PublicClientApplicationBuilder.Create(clientId)
+                                                           .WithAuthority(authority);
+
+            // Listen to MSAL logs if GCM_TRACE_MSAUTH is set
+            if (Context.IsEnvironmentVariableTruthy(Constants.EnvironmentVariables.GcmTraceMsAuth, false))
+            {
+                // If GCM secret tracing is enabled also enable "PII" logging in MSAL
+                bool enablePiiLogging = Context.Trace.IsSecretTracingEnabled;
+
+                appBuilder.WithLogging(OnMsalLogMessage, LogLevel.Verbose, enablePiiLogging, false);
+            }
+
+            IPublicClientApplication app = appBuilder.Build();
+
+            await RegisterVSTokenCacheAsync(app);
 
-            IPlatformParameters parameters = new PlatformParameters(PromptBehavior.SelectAccount);
-            AuthenticationResult result = await authContext.AcquireTokenAsync(
-                resource,
-                clientId,
-                redirectUri,
-                parameters,
-                UserIdentifier.AnyUser);
+            AuthenticationResult result = await app.AcquireTokenInteractive(scopes)
+                                                   .WithPrompt(Prompt.SelectAccount)
+                                                   .ExecuteAsync();
 
             return result.AccessToken;
         }
+
+        private async Task RegisterVSTokenCacheAsync(IPublicClientApplication app)
+        {
+            Context.Trace.WriteLine("Configuring Visual Studio token cache...");
+
+            const string cacheFileName = "msal.cache";
+            string appData = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
+            string cacheDirectory = Path.Combine(appData, ".IdentityService");
+
+            var storageProps = new StorageCreationPropertiesBuilder(cacheFileName, cacheDirectory, app.AppConfig.ClientId).Build();
+
+            var helper = await MsalCacheHelper.CreateAsync(storageProps);
+            helper.RegisterCache(app.UserTokenCache);
+
+            Context.Trace.WriteLine("Visual Studio token cache configured.");
```

Comment:
To make sure I'm on the same page we write the MSAL token to ApplicationData\.IdentityService and VS reads from the same location for it?

Your code is easy to follow but a few comments for newbies up top would be useful.  

Also in the past there have been instances where it's been recommended to delete ADAL token from this location?  Do we think that was correct advice and / or might we have the same requirement with MSAL?

Reply:
> To make sure I'm on the same page we write the MSAL token to ApplicationData.IdentityService and VS reads from the same location for it?

Correct, VS stores the MSAL cache at `%LocalAppData%\.IdentityService\msal.cache`.

> Your code is easy to follow but a few comments for newbies up top would be useful.

Sure!

> Also in the past there have been instances where it's been recommended to delete ADAL token from this location? Do we think that was correct advice and / or might we have the same requirement with MSAL?

We may (in the rarest of circumstances) have that requirement, although I do not think this is desirable nor necessary anymore.